File: Python/canny edge detection.py
# Written by Ivan Reedman
# February 2016
# Canny Edge Detection implementation in Python

# import relevant extensions
from PIL import Image as image
from PIL import ImageDraw
from PIL import ImageFont
import os
import time
import math

# load source image
img_src=image.open("bridge.jpg")

# ensure image is 8bits per pixel
img_src=img_src.convert("L")

# get pixels
pixels_src=img_src.load()

# display image
img_src.show()

# get width and height of source image
width=img_src.size[0]
height=img_src.size[1]

#--------- gaussian blur pass --------
# create destination image for blur
img_blur=image.new("RGB",(width,height),"BLACK")
pixels_blur=img_blur.load()

# get current time
currenttime=time.time()

# display image basics
print "Image width=" + str(width) + " height=" + str(height)
print ""

# display current task
print "Applying gaussian blur to image"

# now calculate gradient intensity of each pixel
for x in range(2,width-2):
    for y in range(2,height-2):

        blurpixel=0

        # +-2 of current position then apply gaussian filter weightings
        for dx in range(-2,2):
            for dy in range(-2,2):
                pixelweight = 0
                if (abs(dy) == 2):
                    if (abs(dx) == 0):
                        pixelweight = 5
                    elif (abs(dx) == 1):
                        pixelweight = 4
                    elif (abs(dx) == 2):
                        pixelweight = 2
                elif (abs(dy) == 1):
                    if (abs(dx) == 0):
                        pixelweight = 12
                    elif (abs(dx) == 1):
                        pixelweight = 9
                    elif (abs(dx) == 2):
                        pixelweight = 4
                elif (abs(dy) == 0):
                    if (abs(dx) == 0):
                        pixelweight = 15
                    elif (abs(dx) == 1):
                        pixelweight = 12
                    elif (abs(dx) == 2):
                        pixelweight = 5

                # apply weighting
                blurpixel = blurpixel + pixels_src[x + dx, y + dy] * pixelweight
        
        # calculate actual result
        blurpixel = blurpixel / 159

        # store gradient into destination image
        pixels_blur[x,y]=(blurpixel,blurpixel,blurpixel)

# display time taken to process image
print "- Apply gaussian blur to image took " + str(round(time.time()-currenttime,3)) + " seconds"
print ""

# copy original to temp
img_temp=img_blur.copy()

# create canvas for drawing
draw = ImageDraw.Draw(img_temp)

# font = ImageFont.truetype(<font-file>, <font-size>)
font = ImageFont.truetype("arial.ttf", 16)

# draw.text((x, y),"message",(r,g,b))
draw.text((0, 0),"Gaussian Blur",(255,0,0),font=font)

# display gradient image and save
img_temp.show()
img_temp.save("blur.bmp","BMP")

# display current task
print "Gradient intensity pass"

#--------- gradient intensity and direction pass --------
# create destination image or black and white
img_grad=image.new("RGB",(width,height),"BLACK")
pixels_grad=img_grad.load()

# create image array for gradient direction
img_grad_dir=image.new("RGB",(width,height),"BLACK")
pixels_grad_dir=img_grad_dir.load()

# get current time
currenttime=time.time()

pix=[0,0,0]

# now calculate gradient intensity of each pixel
for x in range(1,width):
    for y in range(1,height):

        # get source pixels to calculate intensity and direction
        pix[0]=pixels_blur[x,y][0]      # main pixel
        pix[1]=pixels_blur[x-1,y][0]    # pixel left
        pix[2]=pixels_blur[x,y-1][0]    # pixel above

        # get value for x gradient
        grad_x=pix[0]-pix[1]

        # get value for y gradient
        grad_y=pix[0]-pix[2]

        # calculation gradient direction
        # we want this rounded to 0,1,2,3 which represent 0,45,90,135 degrees
        grad_dir = int(abs((math.atan2(grad_y,grad_x))+0.22))*80    # mul by 80 just so image has depth

        # store gradient direction
        pixels_grad_dir[x,y]=(grad_dir,grad_dir,grad_dir)

        # get absolute value for x gradient
        grad_x=abs(grad_x)
        
        # get absolute value for y gradient
        grad_y=abs(grad_y)
        
        # calculate gradient
        grad=int(math.sqrt(grad_x*grad_x+grad_y*grad_y)) * 2

        # store gradient into destination image
        pixels_grad[x,y]=(grad,grad,grad)

# display time taken to process image
print "- Calculating the intensity gradient of image took " + str(round(time.time()-currenttime,3)) + " seconds"
print ""

# copy original to temp
img_temp=img_grad.copy()

# create canvas for drawing
draw = ImageDraw.Draw(img_temp)

# draw.text((x, y),"message",(r,g,b))
draw.text((0, 0),"Gradient",(255,0,0),font=font)

# display gradient image and save
img_temp.show()
img_temp.save("gradient.bmp","BMP")

# copy original to temp
img_temp=img_grad_dir.copy()

# create canvas for drawing
draw = ImageDraw.Draw(img_temp)

# draw.text((x, y),"message",(r,g,b))
draw.text((0, 0),"Gradient Direction",(255,0,0),font=font)

# display gradient direction and save
img_temp.show()
img_temp.save("direction.bmp","BMP")

# display current task
print "Non-maximal suppression"

#--------- non-maximal suppression pass --------
# create destination image or black and white
img_nm=image.new("RGB",(width,height),"BLACK")
pixels_nm=img_nm.load()

# get current time
currenttime=time.time()

# set sensitivity
sensitivity=10

# now suppress non local maximal of each pixel based on direction
for x in range(1,width-1):
    for y in range(1,height-1):
        # is current pixel_grad above threshold
        if (pixels_grad[x,y][0]>=sensitivity):
            # based on direction
            grad_dir=pixels_grad_dir[x,y][0]    # divide by 80 as we multiplied it up just so image was visible

            # if 0 then vertical
            if (grad_dir==0):
                if (pixels_grad[x,y][0]>=pixels_grad[x,y-1][0] and pixels_grad[x,y][0]>= pixels_grad[x,y+1][0]):
                    # mark current pixel as maximal
                    pixels_nm[x,y]=(255,255,255)

                    # mark pixels either side based on direction as non-maximal
                    pixels_nm[x,y-1]=(0,0,0)
                    pixels_nm[x,y+1]=(0,0,0)
                else:
                    # mark current pixel as non-maximal
                    pixels_nm[x,y]=(0,0,0)
                
            # if 80 then 45 degrees
            elif (grad_dir==80):
                if (pixels_grad[x,y][0]>=pixels_grad[x+1,y-1][0] and pixels_grad[x,y][0]>= pixels_grad[x-1,y+1][0]):
                    # mark current pixel as maximal
                    pixels_nm[x,y]=(255,255,255)

                    # mark pixels either side based on direction as non-maximal
                    pixels_nm[x+1,y-1]=(0,0,0)
                    pixels_nm[x-1,y+1]=(0,0,0)
                else:
                    # mark current pixel as non-maximal
                    pixels_nm[x,y]=(0,0,0)

            # if 160 then 90 degrees
            elif (grad_dir==160):
                if (pixels_grad[x,y][0]>=pixels_grad[x-1,y][0] and pixels_grad[x,y][0]>= pixels_grad[x+1,y][0]):
                    # mark current pixel as maximal
                    pixels_nm[x,y]=(255,255,255)

                    # mark pixels either side based on direction as non-maximal
                    pixels_nm[x-1,y]=(0,0,0)
                    pixels_nm[x+1,y]=(0,0,0)
                else:
                    # mark current pixel as non-maximal
                    pixels_nm[x,y]=(0,0,0)

            # if 240 then 135 degrees
            elif (grad_dir==240):
                if (pixels_grad[x,y][0]>=pixels_grad[x-1,y-1][0] and pixels_grad[x,y][0]>= pixels_grad[x-1,y-1][0]):
                    # mark current pixel as maximal
                    pixels_nm[x,y]=(255,255,255)

                    # mark pixels either side based on direction as non-maximal
                    pixels_nm[x-1,y-1]=(0,0,0)
                    pixels_nm[x+1,y+1]=(0,0,0)
                else:
                    # mark current pixel as non-maximal
                    pixels_nm[x,y]=(0,0,0)
        else:
            # mark current pixel as non-maximal
            pixels_nm[x,y]=(0,0,0)
                
# display time taken to process image
print "- Calculating the non-maximal suppression pass of the image took " + str(round(time.time()-currenttime,3)) + " seconds"

# copy original to temp
img_temp=img_nm.copy()

# create canvas for drawing
draw = ImageDraw.Draw(img_temp)

# draw.text((x, y),"message",(r,g,b))
draw.text((0, 0),"Non-maximal suppression",(255,0,0),font=font)

# display nm image
img_temp.show()
img_temp.save("non-maximal.bmp","BMP")

File: C++/Canny Edge Detection.cpp
// Written by Ivan Reedman
// (c) Copyright Ivan J Reedman 2016
// February 2016
// Canny Edge Detection implementation in C++
// Requires CIMG

#include "include\CImg-1.6.9\CImg.h"

using namespace cimg_library;
using namespace std;

#include <iostream>
#include <fstream>
#include <ctime>
#include <string>
#include <cmath>

//using namespace std;

int main()
{
	// Read image 
	CImg<unsigned char> imgsrc("bridge.bmp"); 

	// Display source image
	CImgDisplay main_disp(imgsrc, "Source"); 

	// Define new source image for use
	CImg<unsigned char> imgsrcgrey;
	
	// Convert source RGB into 8bit greyscale image
	imgsrcgrey=imgsrc.get_RGBtoYCbCr().get_channel(0);

	// Display grey scale source image
	CImgDisplay main_disp_grey(imgsrcgrey, "Grey"); 

	// Get width and height of source image
	unsigned int width=imgsrc.width();
	unsigned int height=imgsrc.height();

	// Define standard x and y used throughout code
	unsigned int x=0, y=0;

	// Define standard pixel used throughout code
	unsigned char pixel;

	// Define standard timing vars
	double currenttime;
	double timetaken;

	// Output to console the image basics
	string imagespecifications;
	imagespecifications = "Image Width=" + to_string(width) + " Height=" + to_string(height) + "\r\n\r\n";
	cout << imagespecifications;

	//---------------------------------------------------------------------------------------------
	// --------------------- Gaussian blur -----------------------
	// Define image to store blur
	CImg<unsigned char> imgblur(width, height, 1, 1, 1);
	
	// Get current time
	currenttime = (double)clock();

	// Apply Gaussian Blur
	cout << "Applying gaussian blur to image\r\n";

	// Definitions
	unsigned int blurpixel=0;
	signed int dx = 0, dy = 0;
	unsigned int pixelweight = 0;

	// Define gaussian blur weightings array
	int weighting[5][5] =
	{
		{ 2, 4, 5, 4, 2},
		{ 4, 9,12, 9, 4},
		{ 5,12,15,12, 5},
		{ 4, 9,12, 9, 4},
		{ 2, 4, 5, 4, 2}
	};


	// Get each pixel and apply the blur filter
	for (x = 2; x <= width - 2; x++) {
		for (y = 2; y <= height - 2; y++) {

			// Clear blurpixel
			blurpixel = 0;

			// +- 2 for each pixel and calculate the weighting
			for (dx = -2; dx <= 2; dx++) {
				for (dy = -2; dy <= 2; dy++) {
					pixelweight = weighting[dx+2][dy+2];		

					// Get pixel
					pixel = imgsrcgrey(x + dx, y + dy, 0, 0);

					// Apply weighting
					blurpixel = blurpixel + pixel * pixelweight;
				}
			}

			// Write pixel to blur image
			imgblur(x, y, 0, 0) = (int)(blurpixel / 159);

		}
	}

	// Display time taken to process image
	timetaken = (double)((double)clock() - currenttime) / CLOCKS_PER_SEC;

	std::cout << "- Applying gaussian blur to image took " << timetaken << " seconds\r\n\r\n";
	
	// Display blur image
	CImgDisplay main_disp_blur(imgblur, "blur");

	//---------------------------------------------------------------------------------------------
	// --------------------- Gradient intensity pass -----------------------
	// Define image to store gradient intensity
	CImg<unsigned char> imggrad(width, height, 1, 1, 1);

	// Define image to store gradient direction
	CImg<unsigned char> imggraddir(width, height, 1, 1, 1);

	// Get current time
	currenttime = (double)clock();
	
	// Apply Gaussian Blur
	cout << "Gradient intensity pass\r\n";

	// Definitions
	int pix[3] = { 0,0,0 };
	int gradx = 0, grady = 0;
	int graddir = 0, grad = 0;

	// Get pixels and calculate gradient and direction
	for (x = 1; x <= width-1; x++) {
		for (y = 1; y <= height-1; y++) {

			// Get source pixels to calculate the intensity and direction
			pix[0] = imgblur(x, y, 0, 0);			// main pixel
			pix[1] = imgblur(x - 1, y, 0, 0);		// pixel left
			pix[2] = imgblur(x, y - 1, 0, 0);		// pixel above

			// get value for x gradient
			gradx = pix[0] - pix[1];

			// get value for y gradient
			grady = pix[0] - pix[2];

			// Calculate gradient direction
			// We want this rounded to 0,1,2,3 which represents 0, 45, 90, 135 degrees
			graddir = (int)(abs(atan2(grady, gradx)) + 0.22) * 80;

			// Store gradient direction
			imggraddir(x, y, 0, 0) = graddir;

			// Calculate gradient
			grad = (int)sqrt(gradx * gradx + grady * grady) * 2;

			// Store pixel
			imggrad(x, y, 0, 0) = grad;

		}
	}

	// Display time taken to process image
	timetaken = (double)((double)clock() - currenttime) / CLOCKS_PER_SEC;

	cout << "- Calculating the intensity gradient of the image took " << timetaken << " seconds\r\n\r\n";

	// Display gradient intensity image
	CImgDisplay main_disp_grad(imggrad, "grad");

	// Display gradient direction image
	CImgDisplay main_disp_grad_dir(imggraddir, "graddir");

	//---------------------------------------------------------------------------------------------
	// --------------------- Gaussian blur -----------------------
	// Define image to store blur
	CImg<unsigned char> imgnm(width, height, 1, 1, 1);

	// Get current time
	currenttime = (double)clock(); 
	
	// Apply Gaussian Blur
	cout << "Non-maximal supression pass\r\n";

	// Definitions
	int sensitivity = 10;

	// Get each pixel and apply the blur filter
	for (x = 2; x <= width - 2; x++) {
		for (y = 2; y <= height - 2; y++) {

			// First check that current pixel is above threshold
			if (imggrad(x, y, 0, 0) >= sensitivity) {

				// Get gradient direction
				graddir = imggraddir(x, y, 0, 0);		// Remember this was multiplied by 80 for the sake of display

				// If angle = 0
				if (graddir == 0) {
					// Is pixel local maximal
					if (imggrad(x, y, 0, 0) >= imggrad(x, y - 1, 0, 0) && imggrad(x, y, 0, 0) >= imggrad(x, y + 1, 0, 0)) {
						// Write pixel to as max
						imgnm(x, y, 0, 0) = 255;

						// Supress other two
						imgnm(x, y - 1, 0, 0) = 0;
						imgnm(x, y + 1, 0, 0) = 0;

					}
					else {
						// Supress pixel
						imgnm(x, y, 0, 0) = 0;
					}
				} 

				// If angle = 45 degrees
				else if (graddir == 80) {
					// Is pixel local maximal
					if (imggrad(x, y, 0, 0) >= imggrad(x + 1, y - 1, 0, 0) && imggrad(x, y, 0, 0) >= imggrad(x - 1, y + 1, 0, 0)) {
						// Write pixel to as max
						imgnm(x, y, 0, 0) = 255;

						// Supress other two
						imgnm(x + 1, y - 1, 0, 0) = 0;
						imgnm(x - 1, y + 1, 0, 0) = 0;

					}
					else {
						// Supress pixel
						imgnm(x, y, 0, 0) = 0;
					}
				}

				// If angle = 90 degrees
				else if (graddir == 160) {
					// Is pixel local maximal
					if (imggrad(x, y, 0, 0) >= imggrad(x - 1, y, 0, 0) && imggrad(x, y, 0, 0) >= imggrad(x + 1, y, 0, 0)) {
						// Write pixel to as max
						imgnm(x, y, 0, 0) = 255;

						// Supress other two
						imgnm(x - 1, y, 0, 0) = 0;
						imgnm(x + 1, y, 0, 0) = 0;

					}
					else {
						// Supress pixel
						imgnm(x, y, 0, 0) = 0;
					}
				}

				// If angle = 135 degrees
				else if (graddir == 240) {
					// Is pixel local maximal
					if (imggrad(x, y, 0, 0) >= imggrad(x - 1, y - 1, 0, 0) && imggrad(x, y, 0, 0) >= imggrad(x + 1, y + 1, 0, 0)) {
						// Write pixel to as max
						imgnm(x, y, 0, 0) = 255;

						// Supress other two
						imgnm(x - 1, y - 1, 0, 0) = 0;
						imgnm(x + 1, y + 1, 0, 0) = 0;

					}
					else {
						// Supress pixel
						imgnm(x, y, 0, 0) = 0;
					}
				}
				
			}
			else {

				// Supress pixel
				imgnm(x, y, 0, 0) = 0;

			}

		}
	}

	// Display time taken to process image
	timetaken = (double)((double)clock() - currenttime) / CLOCKS_PER_SEC;

	std::cout << "- Applying non-maxmimal supression to image took " << timetaken << " seconds\r\n";

	// Display blur image
	CImgDisplay main_disp_nm(imgnm, "nm");


	while (!main_disp.is_closed())
		main_disp.wait();
	return 0;
}
